contributors/design-proposals/api-machinery/customresource-conversion-webhook.md

```diff
@@ -0,0 +1,861 @@
+
+* As an author of a CRD that does not use validation, I can still have different versions which undergo conversion.
+
+* As a user, when I request an object, and webhook-conversion fails, I get an error message that helps me understand the problem.
+
+* As an API machinery code maintainer, this change should not make the API machinery code harder to maintain
+
+* As a cluster owner, when I upgrade to the version of Kubernetes that supports CRD multiple versions, but I don't use the new feature, my existing CRDs work fine.  I can roll back to the previous version without any special action.
+
+
+# Summary of Changes
+
+1. A CRD object now represents a group/kind with one or more versions.
+
+2. The CRD API (CustomResourceDefinitionSpec) is extended as follows:
+
+    1. It has a place to register 1 webhook.
+
+    2. it holds multiple "versions".
+
+    3. Some fields which were part of the .spec are now per-version; namely Schema, Subresources, and AdditionalPrinterColumns.
+
+3. A Webhook Host is used to do conversion for a CRD.
+
+    4. CRD authors will need to write a Webhook Host that accepts any version and returns any version.
+
+    5. Toolkits like kube-builder and operator-sdk are expected to provide flows to assist users to generate Webhook Hosts.
+
+
+# Detailed Design
+
+
+## CRD API Changes
+
+The CustomResourceDefinitionSpec is extended to have a new section where webhooks are defined: 
+
+```golang
+// CustomResourceDefinitionSpec describes how a user wants their resource to appear
+type CustomResourceDefinitionSpec struct {
+  Group string
+  Version string
+  Names CustomResourceDefinitionNames
+  Scope ResourceScope
+  // This optional and correspond to the first version in the versions list
+  Validation *CustomResourceValidation
+  // Optional, correspond to the first version in the versions list 
+  Subresources *CustomResourceSubresources
+  Versions []CustomResourceDefinitionVersion
+  // Optional, and correspond to the first version in the versions list  
+  AdditionalPrinterColumns []CustomResourceColumnDefinition
+
+  Conversion *CustomResourceConversion
+}
+
+type CustomResourceDefinitionVersion struct {
+  Name string
+  Served Boolean
+  Storage Boolean
+  // These three fields should not be set for first item in Versions list
+  Schema *JSONSchemaProp
+  Subresources *CustomResourceSubresources
+  AdditionalPrinterColumns []CustomResourceColumnDefinition
+}
+
+Type CustomResourceConversion struct {
+  // Conversion strategy, either "nop” or "webhook”. If webhook is set, Webhook field is required.
+  Strategy string
+
+  // Additional information for external conversion if strategy is set to external
+  // +optional
+  Webhook *CustomResourceConversionWebhook
+}
+
+type CustomResourceConversionWebhook {
+  // ClientConfig defines how to communicate with the webhook. This is the same config used for validating/mutating webhooks.
+  ClientConfig WebhookClientConfig
```

Comment:
original comment by @deads2k asking advice from @liggitt:
> I can't think of a reason for a conversion webhook to require authentication to keep the information a secret, but it may be useful for managing flow control concerns. @liggitt  thoughts? If we need to do it, we should build it in now.